### PR TITLE
duckduckgotool added

### PR DESCRIPTION
**Pull request by Maverick-F35**
### Description
Duck Duck Go Search tool - With this feature, the user can simply get latest information regarding anything in the world and the agent will be able to browse the internet to get that information for them.

### Test Plan
pytests have been structured as functions that start with test_ for all the functions that I have written. They will be found in the test folder.

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update

### Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have read the contributing guide and my code conforms to the guidelines.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have added the required tests.

### CODE REVIEW

File: tests/unit_tests/tools/duck_duck_go/test_duckduckgo_results.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import unittest
+from superagi.tools.duck_duck_go.duck_duck_go_search import DuckDuckGoSearchTool
+
+class DuckDuckGoSearchToolTest(unittest.TestCase):
```
Comment (TransformerOptimus) [inline tests/unit_tests/tools/duck_duck_go/test_duckduckgo_results.py:4]: use pytest for tests dont use unittest.

File: tests/unit_tests/tools/duck_duck_go/test_duckduckgo_toolkit.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import unittest
+from superagi.tools.duck_duck_go.duck_duck_go_search_toolkit import DuckDuckGoToolkit
+from superagi.tools.duck_duck_go.duck_duck_go_search import DuckDuckGoSearchTool
+
+class DuckDuckGoSearchToolKitTest(unittest.TestCase):
```
Comment (TransformerOptimus) [inline tests/unit_tests/tools/duck_duck_go/test_duckduckgo_toolkit.py:5]: use pyttest for tests

File: superagi/tools/DuckDuckGoTool/tests/test_duck_duck_go_search.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/tools/DuckDuckGoTool/tests/test_duck_duck_go_search.py]: Added unit tests inside tests/unittests

File: superagi/tools/DuckDuckGoTool/.gitignore [hunk 4]
```diff
@@ -0,0 +1,160 @@
+# Byte-compiled / optimized / DLL files
+__pycache__/
+*.py[cod]
+*$py.class
+
+# C extensions
+*.so
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/tools/DuckDuckGoTool/.gitignore]: removed .gitignore

File: superagi/tools/DuckDuckGoTool/LICENSE [hunk 5]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/tools/DuckDuckGoTool/LICENSE]: removed

File: superagi/tools/DuckDuckGoTool/duck_duck_go_search.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+        for result in search_results:                                                       #populating the 3 arrays,Each item of the search_result array is an object with keys->"title","href","body"
+            snippets.append(result["title"])  
+            links.append(result["href"])
+            info.append(result["body"])
+
+        webpages=[]                                                                         #webpages array for storing the contents extracted from the links
+        
+        if links:
+            for i in range(0, 3):                                                           #using first 3 links
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/tools/DuckDuckGoTool/duck_duck_go_search.py]: Added constant variables

File: superagi/tools/DuckDuckGoTool/duck_duck_go_search.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            search_results = list(islice(results, 10))                                      #gets first 10 results from results and stores them in search_results
+            if search_results:                                                              #if search result is populated,break as there is no need to attempt the search again
+                break
+
+            time.sleep(1)
+            attempts += 1
+
+        links=[]                                                                            #array for storing all the links generated
+        info=[]                                                                             #array for storing all the information generated
+        snippets=[]                                                                         #array for storing all the snippets/titles generated
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/tools/DuckDuckGoTool/duck_duck_go_search.py]: modified and removed

File: superagi/tools/DuckDuckGoTool/duck_duck_go_search.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/tools/DuckDuckGoTool/duck_duck_go_search.py]: Extracted into methods

File: superagi/tools/DuckDuckGoTool/duck_duck_go_search.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
+import json
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/tools/DuckDuckGoTool/duck_duck_go_search.py]: added

File: tests/unit_tests/tools/duck_duck_go/test_duckduckgo_results.py (hunk 1 above)
Comment (Maverick-F35) [reply tests/unit_tests/tools/duck_duck_go/test_duckduckgo_results.py]: Removed unit tests and added pytest

File: tests/unit_tests/tools/duck_duck_go/test_duckduckgo_toolkit.py (hunk 2 above)
Comment (Maverick-F35) [reply tests/unit_tests/tools/duck_duck_go/test_duckduckgo_toolkit.py]: Added pytests

File: superagi/tools/duck_duck_go/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,4 @@
+pydantic==1.10.8
```
Comment (TransformerOptimus) [inline superagi/tools/duck_duck_go/requirements.txt:1]: this should go to main requirements. Why it is here? only duckduckgo-search is required in main.

Comment (TransformerOptimus) [inline superagi/tools/duck_duck_go/requirements.txt:1]: this file is not required

Comment (Maverick-F35) [reply superagi/tools/duck_duck_go/requirements.txt]: removed

**Review by luciferlinx101 — APPROVED** [review]

**Pull request merged**